### PR TITLE
Add investment function.

### PR DESCRIPTION
**Pull request by songrgg**
1. add and configure EvaTrade module to support VFinance's trade creation and payment.
2. combine financial products and EvaPay to implement the investment workflow.

### CODE REVIEW

File: apps/FMarketApiVer1/src/FMarketApiVer1/Models/Trade.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        $orderEntity->capitalType = Order::TYPE_REAL;
+        $orderEntity->createTime = time();
+        $orderEntity->expiryTime = $expiryTime;
+        $orderEntity->lastModifyTime = $orderEntity->createTime;
+        $orderEntity->type = Order::BUY_PRODUCT;
+        $orderEntity->status = Order::STATUS_PENDING;
+        $orderEntity->investMethod = Order::MANUAL;
+
+        while (!$orderEntity->save()) {
+            foreach ($orderEntity->getMessages() as $message) {
```
Comment (AlloVince) [inline apps/FMarketApiVer1/src/FMarketApiVer1/Models/Trade.php]: 用while太危险

File: apps/FMarketApiVer1/src/FMarketApiVer1/Models/Trade.php [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+     */
+    public static function invest($userId, $mobile, $productId, $capitalAmount, $autopay=false)
+    {
+        $arrProduct = Product::getProduct($productId);
+        if ($arrProduct === false) {
+            throw new Exception\ResourceNotFoundException('Financial product not found');
+        }
+
+        // check the product status
+        $status = isset($arrProduct['status']) ? $arrProduct['status'] : '';
```
Comment (AlloVince) [inline apps/FMarketApiVer1/src/FMarketApiVer1/Models/Trade.php]: 初始化赋值可以用array_merge

File: apps/FMarketApiVer1/src/FMarketApiVer1/Models/Trade.php [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            throw new Exception\InvalidArgumentException('tradeAmount can not be more than the left amount');
+        }
+
+        $expiryTime = isset($arrProduct['expiryTime']) ? $arrProduct['expiryTime'] : 0;
+        $provider = isset($arrProduct['provider']) ? $arrProduct['provider'] : self::WALLSTREETCN;
+        $productName = isset($arrProduct['name']) ? $arrProduct['name'] : '';
+
+        $requestNo = self::getNewOrderNo();
+        $orderEntity = new Order();
+        $orderEntity->vfinanceId = 0;
```
Comment (AlloVince) [inline apps/FMarketApiVer1/src/FMarketApiVer1/Models/Trade.php]: entity绑定值最好用entity->assign()

File: apps/FMarketApiVer1/src/FMarketApiVer1/Models/Trade.php [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+use Eva\EvaEngine\Exception;
+use Eva\EvaEngine\IoC;
+use Eva\EvaPay\Models\VFWallet;
+use Wscn\WscnFMarket\Entities\Order;
+use Wscn\WscnFMarket\Models\Product;
+
+class Trade
+{
+    const WALLSTREETCN = 'WALLSTREETCN';
+    const WALLSTREETCN001 = 'WALLSTREETCN001';
```
Comment (AlloVince) [inline apps/FMarketApiVer1/src/FMarketApiVer1/Models/Trade.php]: 这两个常量意义不明

File: modules/EvaPay/src/EvaPay/Models/VFWallet.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            $inputData[self::FIELD_GO_CACHIER] = 'N';
+            $inputData[self::FIELD_IS_WEB_ACCESS] = 'N';
+        }
+
+        $trade = new WalletTrade([
+            'baseUrl' => $config[self::CONFIG_BASE_URL],
+        ]);
+
+        $result = $trade->create($inputData);
+        if ($result !== false) {
```
Comment (AlloVince) [inline modules/EvaPay/src/EvaPay/Models/VFWallet.php]: 底层调CURL的话必须设置超时

File: modules/WscnFMarket/src/WscnFMarket/Models/Product.php [hunk 6]
```diff
@@ -1,13 +1,25 @@
...
-use Wscn\WscnFMarket\Entities\Products;
-
-class ProductModel
+class Product extends \Wscn\WscnFMarket\Entities\Products
 {
     const COLUMN_TAGS = 'tags';
 
+    public static function getProduct($productId)
+    {
+        $product = self::findFirst(["id=$productId"]);
```
Comment (AlloVince) [inline modules/WscnFMarket/src/WscnFMarket/Models/Product.php]: 这里有SQL注入风险

File: apps/FMarketApiVer1/src/FMarketApiVer1/Models/Trade.php [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    const RANDOM_MIN = 10000000000000;
+    const RANDOM_MAX = 90000000000000;
+
+    private static function getNewOrderNo()
+    {
+        return strftime('%Y%m%d%H%M%S') . rand(self::RANDOM_MIN, self::RANDOM_MAX);
```
Comment (chaiyangyang) [inline apps/FMarketApiVer1/src/FMarketApiVer1/Models/Trade.php]: 备注一下是外部订单号。
hash一下使它更没意义。

File: apps/FMarketApiVer1/src/FMarketApiVer1/Models/Trade.php [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+        // check the minimum investment amount
+        $baseAmount = isset($arrProduct['baseAmount']) ? $arrProduct['baseAmount'] : 0;
+        if ($baseAmount > $capitalAmount) {
+            throw new Exception\InvalidArgumentException('Trade amount can not be less than ' . $baseAmount);
+        }
+
+        // check if match the increment amount
+        $incrementalAmount = isset($arrProduct['incrementalAmount']) ? $arrProduct['incrementalAmount'] : 0;
+        if ($incrementalAmount > 0 && ($capitalAmount - $baseAmount) % $incrementalAmount > 0) {
```
Comment (chaiyangyang) [inline apps/FMarketApiVer1/src/FMarketApiVer1/Models/Trade.php]: ($capitalAmount - $baseAmount) % $incrementalAmount != 0)

**Pull request merged**